maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java

```diff
@@ -999,6 +1000,9 @@ class InternalConfig {
     }
 
     private ModelCache createModelCache(RepositorySystemSession session) {
-        return modelCacheFactory.createCache(session);
+        // MNG-7693: for older clients (not injecting ModelCacheFactory), make this work OOTB w/ defaults
+        return modelCacheFactory != null
+                ? modelCacheFactory.createCache(session)
+                : new DefaultModelCacheFactory().createCache(session);
```

Comment:
Can we save result to avoid many instances creation?
How often this method is used?

Reply:
We can investigate this, yes, but for this PR I'd really not change any semantics here, just fix the NPE for older clients. ie. restore backward compatibility.